api/pkg/auth/auth.go

```diff
@@ -0,0 +1,151 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/markbates/goth/gothic"
+	"github.com/tektoncd/hub/api/gen/log"
+	"github.com/tektoncd/hub/api/pkg/app"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"gorm.io/gorm"
+)
+
+type service struct {
+	app.Service
+	api app.Config
+}
+
+type request struct {
+	db            *gorm.DB
+	log           *log.Logger
+	defaultScopes []string
+	jwtConfig     *app.JWTConfig
+}
+
+type ProviderList struct {
+	Provider      string
+	EnterpriseUrl string
+	ClientId      string
+}
+
+var (
+	UI_URL string
+)
+
+type Service interface {
+	AuthCallBack(res http.ResponseWriter, req *http.Request)
+	HubAuthenticate(res http.ResponseWriter, req *http.Request)
+}
+
+// New returns the resource service implementation.
+func New(api app.Config) Service {
+	return &service{
+		Service: api.Service("auth"),
+		api:     api,
+	}
+}
+
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	UI_URL = req.FormValue("redirect_uri")
+	gothic.BeginAuthHandler(res, req)
+}
+
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		fmt.Fprintln(res, err)
+		return
+	}
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	params := req.URL.Query()
+
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
+		s.Logger(context.Background()).Error(err)
+		res.Header().Set("Location", fmt.Sprintf("%s?status=%d", UI_URL, http.StatusBadRequest))
+		res.WriteHeader(http.StatusTemporaryRedirect)
+	}
+
+	res.Header().Set("Location", fmt.Sprintf("%s?status=%d&code=%s", UI_URL, http.StatusOK, params.Get("code")))
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+func (s *service) HubAuthenticate(res http.ResponseWriter, req *http.Request) {
+
+	code := req.FormValue("code")
```

Comment:
can we clear this code from db after the login is successful so that nobody can misuse it later on?

Reply:
Yes that can be done 👍🏻

Reply:
will update it